### ISSUE TITLE
Hashing elm/bytes

### ISSUE THREAD

**Issue by zoul** [TSFoster/elm-sha1#5]
The readme says:

> Currently, the package can only create digests for around 200kb of data. If there is any interest in using this package for hashing >200kb, or for hashing elm/bytes, let me know!

I’d like both :) I want to hash some images (megabytes of data) before uploading them somewhere, see [my question at Elm Discourse for details](https://discourse.elm-lang.org/t/how-do-i-compute-the-sha1-sum-of-a-file/4454).

**Comment by TSFoster** [comment]
Hi @zoul,

Cool! I originally wrote this before elm/bytes was published, which is why it hasn't had support for it yet.

I am up for fixing the 200+kb issue (I vaguely recall thinking a certain part could be slightly rewritten to be properly tail-call optimised, or could be converted to a fold), and for having a look at adding elm/bytes support.

Unfortunately, I won't be able to look at it for about 3 weeks, but I've put it in my calendar to have a look then. Otherwise, pull requests are appreciated!

**Comment by folkertdev** [comment]
I've been working on `elm-bytes` support on [this fork](https://github.com/folkertdev/elm-sha1). It's kind of messy right now as I'm experimenting, I'll try to break it up into self-contained commits.

I think some design work is needed though. The `fromBytes` name is kind of confusing when it can mean both a list of byte integer values, and a `Bytes` value. Also I'm not that familiar with the terminology so picking good names for variables/helpers is hard. But I have a lot of experience with making elm/bytes go fast so I'm happy to help out. 

As a side-effect, using bytes is also ~20% faster, and means the dependency on utf-8 tools is not required any more. Some further gains can be made by inlining list operations. (e.g. `sum [1,2,3] is much slower than `1 + 2 + 3` in tight loops)

Finally, the 200kb limitation is because of speed? Kind of makes sense because currently that would mean a 200000 element list which won't be fast. Just want to make sure that the logic in theory should be correct even for larger input sizes.

**Comment by folkertdev** [comment]
an update: 

My version is at least 4x faster, probably much more for larger inputs. I seem to get a stable speed of 5 to 6 Mbps. The tradeoff is that the code is not quite as nice to look at. 

What mainly matters, I think, is that `Bytes` is just so much faster and more compact than a `List Int` when working with byte values. Additionally, the code only makes one traversal of the input and doesn't copy (the current version uses `List.Extra.groups` which duplicates the input in memory).

Another trick is inlining, in particular of an `Array Int` (that would again duplicate the input in memory). The result is functions with 16+ arguments, but they are fast.

These changes are all pretty drastic, so I'm not sure if just one PR is the best way to go. @zoul you should be able to add my version as a source file, some real life speed data would be useful.

**Comment by TSFoster** [comment]
I've just published 2.0.0, which includes @folkertdev's performance improvements and support for elm/bytes, along with the welcome API changes.

I've also published, to reach as many as possible:
* 1.0.5 as a patch release, which introduces the performance improvements
* 1.1.0 as a minor release, which introduces full elm/bytes support without changing the existing API

**Comment by zoul** [comment]
(I’m sorry for not responding sooner – this is perfect, thank you very much.)

**Comment by benkoshy** [comment]
Hi there @TSFoster @folkertdev  - appreciate the work you've put into the library.

### Taking excessively long to calculate SHAs of large files:

I'm trying to use the `SHA1.fromByteValues` method to calculate the SHA1 of a fairly sizable file: some 80 MB but it seems to be taking an excessively long time period to calculate (over 20 seconds):

Here is an Ellie to demonstrate the issue:

https://ellie-app.com/99qN9FsPQv5a1

And here is the offending code, in short:

```elm
        GotFiles file files ->
            ( { model | hover = False }
            , (file :: files)
                |> List.map
                    (File.toBytes
                        >> Task.map (SHA1.fromBytes >> SHA1.toHex)
                    )
                |> Task.sequence
                |> Task.perform GotShas
            )
```
Any pointers in the right direction would be much appreciated.

(Also, thank you so much for your library )!

rgds
Ben

**Comment by TSFoster** [comment]
Hi @BKSpurgeon, thanks for the explanation and brilliant short, self-contained example.

I've never had to calculate the SHA1 of a large file with this package before—I agree, it takes too long!

I imagine there may be a couple of small performance improvements that could be made, but maybe not the scale you're hoping for. I modified your Ellie to time reading the file vs calculating the digest vs performing a single bitwise operation on each byte ("xorAll"): <https://ellie-app.com/99bBskZ3vDJa1>. I tried it with a 90MB file and it took about 1s to just read the file, 8s for "xorAll", and 17s to calculate the SHA1 digest. In contrast, calculating the digest in the terminal with `sha1sum` took 0.2s.

@folkertdev: do you have any ideas for how to try improving performance? Could changing the `Tuple5` type from a record to a Custom Type make a difference? Or changing two `if` statements to `case`s?

**Comment by folkertdev** [comment]
First thing is that ellie does not use `--optimize`. That is probably significant, because elm-sha1 uses a bunch of newtype wrappers internally. Those get optimized away by `--optimize`. So if you could run that ellie locally with `--optimize` I think the performance would be better already. 

I'll make a PR with some further changes that give a bit more speedup. Re. your suggestions

* custom types and records are represented in exactly the same way (a javascript object). So that would not make a difference.
* elm optimizes `if` and `case` heavily, so I don't think that's the problem

**Comment by benkoshy** [comment]
@TSFoster @folkertdev thank you for your reply.

### Could not reproduce Sha1 values calculated

In addition to the above, I was not able to reproduce the SHA1 value calculated vs. using other means to calculated the SHA1  - I've created a repo where this can be tested locally with the exact file that I was using: 

https://github.com/BKSpurgeon/elm-sha-example

Again, thank you for the work you've both put into this library.

**Comment by folkertdev** [comment]
I got a 2X speed improvement and a fix for the incorrect sha1 value in #8.

**Comment by benkoshy** [comment]
awesome work @folkertdev much appreciated.
